sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/PortEvaluationStage.java

```diff
@@ -0,0 +1,125 @@
+                dynamicPort = availablePorts.getRanges().getRangeList().stream()
+                        .flatMap(r -> IntStream.rangeClosed((int) r.getBegin(), (int) r.getEnd()).boxed())
+                        .filter(p -> !consumedPorts.contains(p))
+                        .findFirst();
+            }
+
+            if (!dynamicPort.isPresent()) {
+                throw new OfferEvaluationException(String.format(
+                        "No ports were available for dynamic claim in offer: %s",
+                        offerResourcePool.getOffer().toString()));
+            }
+            assignedPort = dynamicPort.get();
+        }
+
+        super.setResourceRequirement(getDynamicPortRequirement(getResourceRequirement(), assignedPort));
+        super.evaluate(offerResourcePool, offerRequirement, offerRecommendationSlate);
+    }
+
+    @Override
+    protected void setProtos(OfferRequirement offerRequirement, Protos.Resource resource) {
+        long port = resource.getRanges().getRange(0).getBegin();
+
+        if (getTaskName().isPresent()) {
+            String taskName = getTaskName().get();
+            Protos.TaskInfo taskInfo = offerRequirement.getTaskRequirement(taskName).getTaskInfo();
+            Protos.TaskInfo.Builder taskInfoBuilder = taskInfo.toBuilder();
+            Protos.Resource.Builder ports = ResourceUtils.getResource(taskInfo, Constants.PORTS_TYPE).toBuilder();
+
+            ports.getRangesBuilder().addRangeBuilder().setBegin(port).setEnd(port);
+            ResourceUtils.setResource(taskInfoBuilder, ports.build());
+
+            taskInfoBuilder.setCommand(
+                    CommandUtils.addEnvVar(
+                            taskInfoBuilder.getCommand(), getPortEnvironmentVariable(portName), Long.toString(port)));
+            offerRequirement.updateTaskRequirement(taskName, taskInfoBuilder.build());
+        } else {
+            Protos.ExecutorInfo executorInfo = offerRequirement.getExecutorRequirementOptional()
+                    .get()
+                    .getExecutorInfo();
+            Protos.ExecutorInfo.Builder executorInfoBuilder = executorInfo.toBuilder();
+            Protos.Resource.Builder ports = ResourceUtils.getResource(executorInfo, Constants.PORTS_TYPE).toBuilder();
+
+            ports.getRangesBuilder().addRangeBuilder().setBegin(port).setEnd(port);
+            ResourceUtils.setResource(executorInfoBuilder, ports.build());
+            executorInfoBuilder.setCommand(
+                    CommandUtils.addEnvVar(
+                            executorInfoBuilder.getCommand(),
+                            getPortEnvironmentVariable(portName),
+                            Long.toString(port)));
+            offerRequirement.updateExecutorRequirement(executorInfoBuilder.build());
+        }
+    }
+
+    private static String getPortEnvironmentVariable(String portName) {
+        return "PORT_" + portName;
+    }
+
+    private static ResourceRequirement getDynamicPortRequirement(
```

Comment:
This function isn't specific to dynamic ports, appears to be called regardless of dynamic vs fixed?